src/openscm_runner/adapters/fair2/fair2.py

```diff
@@ -0,0 +1,316 @@
+
+    if scmdf[["model", "scenario"]].drop_duplicates().shape[0] != 1:
+        raise AssertionError("Should only have one model-scenario pair")
+
+    scen_start_year = 2015
+
+    scmdf = ScmRun(scmdf.timeseries()).interpolate(
+        [dt.datetime(y, 1, 1) for y in range(scen_start_year, endyear + 1)]
+    )
+
+    years = scmdf["year"].values
+    first_scenyear = years[0]
+    first_scen_row = int(first_scenyear-startyear)
+
+    # if correct units and interpolation were guaranteed we could do this for scenario too which is quicker
+    hist_df = ssp245_world_emms_holder.values_fair_units.filter(
+        year=range(startyear, 2015)
+    ).timeseries()
+
+    future_ssp245_df = ssp245_world_emms_holder.values_fair_units.filter(
+        year=range(2015, endyear + 1)
+    ).timeseries()
+
+    for species in EMISSIONS_SPECIES_UNITS_CONTEXT["species"]:
+        fair_col, _, _ = _get_fair_col_unit_context(species)
+
+        hist_df_row = hist_df.index.get_level_values("variable").str.endswith(species)
+
+        data_out[: first_scen_row, fair_col] = hist_df[hist_df_row].values.squeeze()
+
+        future_ssp245_df_row = future_ssp245_df.index.get_level_values("variable").str.endswith(species)
+
+        data_out[first_scen_row :, fair_col] = future_ssp245_df[future_ssp245_df_row].values.squeeze()
+
+
+    for var_df in scmdf.groupby("variable"):
+        variable = var_df.get_unique_meta("variable", no_duplicates=True)
+        in_unit = var_df.get_unique_meta("unit", no_duplicates=True)
+        fair_col, fair_unit, context = _get_fair_col_unit_context(variable)
+
+        if in_unit != fair_unit:
+            var_df_fair_unit = var_df.convert_unit(fair_unit, context=context)
+        else:
+            var_df_fair_unit = var_df
+
+        data_out[first_scen_row :, fair_col] = var_df_fair_unit.values.squeeze()
+
+    df_out = pd.DataFrame(  index = data_out[:,0],\
+                            columns = EMISSIONS_SPECIES_UNITS_CONTEXT["species"],\
+                            data = data_out[:,1:])
+
+    return df_out
+
+class FAIR2(_Adapter):
+    """
+    Adapter for running FAIR
+    """
+
+    def _init_model(self, *args, **kwargs):
+        pass
+
+    def _run(self, scenarios, cfgs, output_variables, output_config):
+        if output_config is not None:
+            raise NotImplementedError("`output_config` not implemented for FaIR")
+
+        fair_df = ScmRun(scenarios.timeseries())
+        full_cfgs = self._make_full_cfgs(fair_df, cfgs)
+
+        res = run_fair(full_cfgs, output_variables)
+        res["climate_model"] = "FaIRv{}".format(self.get_version())
+
+        return res
+
+    def _make_full_cfgs(self, scenarios, cfgs):  # pylint: disable=R0201
+        full_cfgs = []
+        run_id_block = 0
+
+        for (scenario, model), smdf in tqdm(
+            scenarios.timeseries().groupby(["scenario", "model"]),
+            desc="Creating FaIR emissions inputs",
+        ):
+            smdf_in = ScmRun(smdf)
```

Comment:
can be removed if the change above is made

Reply:
Done, this makes sense - why bother going scmdf -> scm_time_series_df -> scmdf